### DISCUSSION TITLE
Homework 8: Amount sell

### DISCUSSION THREAD

**Discussion by Di3goCH** [cafawo/FinancialDataAnalytics#17] [Q&A] [answered]
Hi all, 

I wanted to share what I found out about the error in the selling code in the last class. 
I found that the amount has to be a multiple of 10 because a contract in Deribit has a minimum value of 10 USD. That is why sometimes the code worked and sometimes not. 

I change the code in amount, now it works for me always :

```python
params = {
    "amount":  int(round(market_price/10,0)*10), 
    "instrument_name": "BTC-PERPETUAL",
    "type": "market"
    } 
```

**Comment by CS-MLU** [comment]
The solution works for me in the first task but as the the second task asks for only 60%/50% of the price I seem to run into the same issue as before.

**Reply by Di3goCH** [reply]
Could you share the Code?

**Reply by CS-MLU** [reply]
#define the URL of the endpoint for selling #Task 2
url = "https://test.deribit.com/api/v2/private/sell"

#define the parameters 
params = {
    "amount": int(round(market_price/10,0)*6), #60%
    "instrument_name": "BTC-PERPETUAL",
    "label": "test02",
    "type": "stop_limit",
    "time_in_force": "good_til_day",
    "trigger_price": int(round(market_price/10,0)*5), #50%
    }


#send the GET request
response = requests.get(url, params=params, auth=(ClientID, ClientSecret))
response_json = response.json()
response_json['result']

**Reply by Di3goCH** [reply]
I think it is because the amount that you get is not a multiple of 10, try it with this code instead. "int(round(market_price*0.6/10)*10)"

**Reply by CS-MLU** [reply]
Unfortunately that does not work

**Reply by Di3goCH** [reply]
And what is the error message from response_json['result']?

**Reply by CS-MLU** [reply]
Yeah, its the same result

**Reply by Di3goCH** [reply]
@CS-MLU I think you need to add "price" in the parameters. It says it is not required, but later it says that it is only for stop and stop_limit orders.

also "trigger_price": int(round(market_price/10,0)*5), #50% 
should be:
 "trigger_price": int(round(market_price*0.5/10))*10, #50%

**Comment by OlgaKuzminichna** [comment]
I have the same issue: here is my version:

response_json = response.json()
market_price = response_json["result"]["trades"][0]["mark_price"]
print(market_price)
#Define the URL of the endpoint for selling
url = "https://test.deribit.com/api/v2/private/sell"
params = {
    "amount": int(round(market_price / 10, 0) * 10),
    "instrument_name": "BTC-PERPETUAL",
    "type": "stop_limit",
     "stop_price": round(market_price * 0.5), 
     "price": round(market_price * 0.6),  
    "time_in_force": "good_til_day"  
}

#Send the GET request
response = requests.get(url, params=params, auth=(ClientID, ClientSecret))

#Check if the request was successful
if response.status_code == 200:
    response_json = response.json()
    result = response_json['result']
    print(json.dumps(result, indent=4))
else:
    print(f"Failed to retrieve data: {response.status_code}")

**Reply by Di3goCH** [reply]
@olyafrontend I think instead of stop_price you have to use trigger_price. In the documentation I cannot find stop_price for private_sell.

**Comment by Di3goCH** [comment]
I can share with you what I did: 

Function Stop_limit_order () inside a class MarketHandler:

```python
    def Stop_limit_order(self,amount,instrument_name, time_in_force,price, trigger_price):
         
        url = "https://test.deribit.com/api/v2/private/sell"
        params = {
            "amount": amount, 
            "instrument_name": instrument_name,
            "type": "stop_limit",
            "trigger_price": trigger_price,
            "price": price,
            "time_in_force": time_in_force,
            "trigger":"mark_price"
              }
        response = requests.get(url, params=params, auth=(self.ClientID, self.ClientSecret))

        if response.status_code == 200:
           response = response.json()
           order_id=response["result"]["order"]["order_id"]
           print(f"The order was carried out satisfactorily. Order ID: {order_id}")

        else:
           print(f"Failed to retrieve data: {response.status_code}")
``` 

For use the function I have: 
```python
instrument_name="BTC-PERPETUAL"
price=int(round(Seller.GetMarketPrice(instrument_name=instrument_name)*0.6/10)*10)
trigger_price=int(round(Seller.GetMarketPrice(instrument_name=instrument_name)*0.5/10)*10)
amount=int(price/10)*10
time_in_force="good_til_day"
instrument_name="BTC-PERPETUAL"

Seller.Stop_limit_order(amount=amount,instrument_name=instrument_name,price=price,trigger_price=trigger_price,time_in_force=time_in_force)
```

**Comment by birte0** [comment, marked as answer]
Adding to @Di3goCH  suggestion, one can also use `int(market_price.round(-1))` to round the current market price to the neares ten. 

Of cause, if you then calculate some percentage of the price, first calculate the percentage and then round to the closest ten (as otherwise the stop_price would be for example 26530*0.5=13265, which is not a multiple of ten). Instead use: `int((market_price*0.5).round(-1))`

**Reply by Di3goCH** [reply]
Thanks! That's much shorter :)
